gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+        "Performing lazySeek from %s to %s position for '%s'",
+        contentChannelPosition, currentPosition, resourceId);
+    // record the variation in channelPosition
+    // used to auto-detect random access
+    long oldPosition = contentChannelPosition;
+    long seekDistance = currentPosition - contentChannelPosition;
+    // Check if inPlace seek is a viable option
+    // If not close existing channel and open new one.
+
+    if (contentReadChannel != null
+        && seekDistance > 0
+        && seekDistance <= readOptions.getInplaceSeekLimit()
+        && currentPosition < contentChannelEnd) {
+      skipInPlace(seekDistance);
+    } else {
+      closeContentChannel();
+    }
+    if (contentReadChannel == null) {
+      if (isRandomAccessPattern(oldPosition)) {
+        setRandomAccess();
+      }
+      contentReadChannel = openReadChannel(bytesToRead);
+    }
+  }
+
+  // Update this function to create the reader and then set limits
+  private ReadableByteChannel getStorageReadChannel(long seek, long limit) throws IOException {
+    ReadChannel readChannel = storage.reader(blobId, generateReadOptions(blobId));
+    try {
+      readChannel.seek(seek);
+      // TODO: should we also add a check to not set limit if contentChannelEnd == objectSize?
+      if (limit != -1) {
+        readChannel.limit(limit);
+      }
+      return readChannel;
+    } catch (Exception e) {
+      throw new IOException(
+          String.format(
+              "Unable to update the boundaries/Range of contentChannel %s", resourceId.toString()),
+          e);
+    }
+  }
+
+  private static BlobSourceOption[] generateReadOptions(BlobId blobId) {
+    List<BlobSourceOption> readOptions = new ArrayList<>();
+    if (blobId.getGeneration() != null) {
+      readOptions.add(BlobSourceOption.generationMatch(blobId.getGeneration()));
+    }
+    return readOptions.toArray(new BlobSourceOption[readOptions.size()]);
+  }
+
+  private boolean isFooterRead() {
+    return objectSize - currentPosition <= footerSize;
+  }
+
+  private static void validate(GoogleCloudStorageItemInfo itemInfo) throws IOException {
+    checkNotNull(itemInfo, "itemInfo cannot be null");
+    StorageResourceId resourceId = itemInfo.getResourceId();
+    checkArgument(
+        resourceId.isStorageObject(), "Can not open a non-file object for read: %s", resourceId);
+    if (!itemInfo.exists()) {
+      throw new FileNotFoundException(String.format("File not found: %s", resourceId));
+    }
+    // The non-gRPC read channel has special support for gzip. This channel doesn't
+    // decompress gzip-encoded objects on the fly, so best to fail fast rather than return
+    // gibberish unexpectedly.
+    String contentEncoding = itemInfo.getContentEncoding();
+    if (contentEncoding != null && contentEncoding.contains("gzip")) {
+      throw new IOException(
+          String.format(
+              "Cannot read GZIP-encoded file (%s) (not supported via gRPC API): %s",
```

Comment:
customer do not know about gRPC here correct? This error message can be confusing.

Reply:
May be but it's still required and helpful for us to debug.